ibis/expr/analysis.py

```diff
@@ -1039,75 +1037,65 @@ def is_scalar_reduction(expr):
+            tables=tables,
+            predicates=predicates,
         )
-        self.foreign_table: ir.Table | None = None
-        self.predicates: MutableSequence[ir.BooleanValue] = []
+    return op.to_expr()
 
-    def get_result(self) -> ir.BooleanColumn:
-        self._visit(self.expr)
 
-        op = self.expr.op()
+class _AnyToExistsTransform:
+    __slots__ = "expr", "tables", "predicates"
 
-        if isinstance(op, ops.Any):
-            op_class = ops.ExistsSubquery
-        else:
-            assert isinstance(op, ops.NotAny)
-            op_class = ops.NotExistsSubquery
+    def __init__(self, expr: ir.Expr) -> None:
+        self.expr = expr
+        self.tables: MutableSet[ir.Table] = set(expr.op().root_tables())
+        self.predicates: MutableSequence[ir.BooleanValue] = []
 
-        return op_class(self.foreign_table, self.predicates).to_expr()
+    def get_result(self) -> tuple[Sequence[ir.Table], list[ir.BooleanColumn]]:
+        self._visit(self.expr)
+        return (
+            tuple(table.to_expr() for table in self.tables),
+            self.predicates,
+        )
 
     def _visit(self, expr: ir.Expr) -> None:
-        node = expr.op()
-
+        if isinstance(expr, ir.Table):
+            self._visit_table(expr)
+        elif isinstance(expr, ir.BooleanColumn):
+            for sub_expr in flatten_predicate(expr):
+                self.predicates.append(sub_expr)
+                self._visit_node(sub_expr.op())
+        elif isinstance(expr, ir.Expr):
+            self._visit_node(expr.op())
+
+    def _visit_node(self, node: ops.Node) -> None:
         for arg in node.flat_args():
-            if isinstance(arg, ir.Table):
-                self._visit_table(arg)
-            elif isinstance(arg, ir.BooleanColumn):
-                for sub_expr in flatten_predicate(arg):
-                    self.predicates.append(sub_expr)
-                    self._visit(sub_expr)
-            elif isinstance(arg, ir.Expr):
-                self._visit(arg)
-
-    def _find_blocking_table(self, expr: ir.Expr) -> ir.Expr:
-        node = expr.op()
+            self._visit(arg)
 
-        if node.blocks():
-            return expr
-
-        return next(
-            (
-                result
-                for arg in node.flat_args()
-                if (
-                    isinstance(arg, ir.Expr)
-                    and (result := self._find_blocking_table(arg)) is not None
-                )
-            ),
-            None,
-        )
+    def _visit_table(self, table: ir.Table) -> None:
+        assert isinstance(table, ir.Table)
 
-    def _visit_table(self, expr: ir.Expr) -> None:
-        if (
-            isinstance(expr, ir.Table)
-            and (base_table := self._find_blocking_table(expr)) is not None
-            and base_table.op() not in self.query_roots
-        ):
-            self.foreign_table = expr
+        if any(op.blocks() for op in util.to_op_dag(table)):
```

Comment:
`to_op_dag()` is equivalent with:

```python
def has_blocking_table(expr):
    def predicate(expr):
        return lin.proceed, expr.op()
           
    return lin.traverse(predicate, expr)
```

At some point we should unify these two utility functions.

Reply:
Indeed. Let's do that in a follow-up.